superset-frontend/src/components/TableSelector/index.tsx

```diff
@@ -174,56 +185,73 @@ const TableSelector: FunctionComponent<TableSelectorProps> = ({
     // reset selections
     if (database === undefined) {
       setCurrentDatabase(undefined);
+      setCurrentCatalog(undefined);
       setCurrentSchema(undefined);
       setCurrentTable(undefined);
     }
   }, [database]);
 
   useEffect(() => {
-    if (currentDatabase && currentSchema) {
+    if (shouldLoadTables(currentDatabase)) {
       setLoadingTables(true);
-      const encodedSchema = encodeURIComponent(currentSchema);
       const forceRefresh = refresh !== previousRefresh;
-      // TODO: Would be nice to add pagination in a follow-up. Needs endpoint changes.
-      const endpoint = encodeURI(
-        `/superset/tables/${currentDatabase.id}/${encodedSchema}/undefined/${forceRefresh}/`,
-      );
-
-      if (previousRefresh !== refresh) {
-        setPreviousRefresh(refresh);
-      }
-
-      SupersetClient.get({ endpoint })
-        .then(({ json }) => {
-          const options: TableOption[] = [];
-          let currentTable;
-          json.options.forEach((table: Table) => {
-            const option = {
-              value: table.value,
-              label: <TableOption table={table} />,
-              text: table.label,
-            };
-            options.push(option);
-            if (table.label === tableName) {
-              currentTable = option;
-            }
-          });
-
-          onTablesLoad?.(json.options);
-          setTableOptions(options);
-          setCurrentTable(currentTable);
-          setLoadingTables(false);
-          if (forceRefresh) addSuccessToast('List updated');
-        })
-        .catch(() => {
-          setLoadingTables(false);
-          handleError(t('There was an error loading the tables'));
-        });
+      const endpoint = getEndpoint({
+        forceRefresh,
+        schema: currentSchema || '',
+        databaseId: currentDatabase?.id || 0,
+      });
+      const encodedEndpoint = encodeURI(endpoint);
+      if (previousRefresh !== refresh) setPreviousRefresh(refresh);
+      fetchTables(encodedEndpoint, forceRefresh);
     }
-    // We are using the refresh state to re-trigger the query
-    // previousRefresh should be out of dependencies array
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [currentDatabase, currentSchema, onTablesLoad, refresh]);
+  }, [refresh, currentDatabase, currentCatalog, currentSchema, onTablesLoad]);
+
+  function getEndpoint({
+    schema,
+    databaseId,
+    forceRefresh,
+  }: {
+    schema: string;
+    databaseId: number;
+    forceRefresh: boolean;
+  }) {
+    const encodedSchema = encodeURIComponent(schema);
+    return `/superset/tables/${databaseId}/${encodedSchema}/undefined/${forceRefresh}/`;
+  }
```

Comment:
While we're at it, it would be nice to wrap these in `useCallback`. Also, you should always use arrow functions in React components (applies to `shouldLoadTables` and other functions below).